platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultConfigurationCache.kt

```diff
@@ -283,12 +284,18 @@ class DefaultConfigurationCache internal constructor(
             cacheIO.writeCacheEntryDetailsTo(buildStateRegistry, usedModels, usedMetadata, sideEffects, layout.fileFor(StateType.Entry))
         }
         store.useForStore { layout ->
-            val existingEntries = cacheIO.readCandidateEntries(layout.fileForRead(StateType.Candidates))
-            val newEntries = listOf(CandidateEntry(entryId)) + existingEntries
+            val newEntries = calculateCandidateEntries(layout, startParameter.entriesPerKey)
             cacheIO.writeCandidateEntries(layout.fileFor(StateType.Candidates), newEntries)
         }
     }
 
+    private
+    fun calculateCandidateEntries(layout: ConfigurationCacheRepository.Layout, entriesPerKey: Int): List<CandidateEntry> {
+        val existingEntries = cacheIO.readCandidateEntries(layout.fileForRead(StateType.Candidates))
+        val tail = existingEntries.take(min(existingEntries.size, entriesPerKey - 1))
```

Comment:
💭 I'm not very fond of blindingly discarding the oldest entry when it can actually be the most common. What if we also changed `checkFingerprint` to move the found entry to the front to make it LRU?